alf/utils/conditional_ops.py

```diff
@@ -0,0 +1,100 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Conditional operations."""
+
+import tensorflow as tf
+
+
+def _gather_nest(nest, indices):
+    return tf.nest.map_structure(lambda t: tf.gather(t, indices, axis=0), nest)
+
+
+def select_from_mask(data, mask):
+    """Select the items from data based on mask.
+
+    data[i,...] will be selected to form a new tensor if mask[i] is True
+
+    Args:
+        data (nested Tensor): source tensor
+        mask (Tensor): dtype is tf.bool
+    Returns:
+        nested Tensor with the same structure as data
+    """
+    scatter_indices = tf.where(mask)
+    gather_indices = tf.squeeze(scatter_indices, 1)
+    return _gather_nest(data, gather_indices)
+
+
+def conditional_update(target, cond, func, *args, **kwargs):
+    """Update target according to cond mask
+
+    target[row] is changed to func(*args[row], **kwargs[row]) if cond[row]
+    is True, otherwise target[row] remains unchanged.
+
+    Args:
+        target (nested Tensor): target to be updated
+        func (Callable): a function with arguments (*args, **kwargs) and returning
+            a nest with same structure as target
+        cond (Tensor): 1d bool Tensor with shape[0] == target.shape[0]
+    Returns:
+        nest with the same structure and shape as target.
+    """
+    # shape of indices from where() is [batch_size,1], which is what scatter_nd
+    # needs
+    scatter_indices = tf.where(cond)
+
+    def _update_subset():
+        gather_indices = tf.squeeze(scatter_indices, 1)
+        selected_args = [_gather_nest(arg, gather_indices) for arg in args]
+        selected_kwargs = dict(
+            [(k, _gather_nest(v, gather_indices)) for k, v in kwargs.items()])
+        updates = func(*selected_args, **selected_kwargs)
+        return tf.nest.map_structure(
+            lambda tgt, updt: tf.tensor_scatter_nd_update(
+                tgt, scatter_indices, updt), target, updates)
+
+    total = tf.shape(cond)[0]
+    n = tf.shape(scatter_indices)[0]
+    return tf.case(((n == 0, lambda: target),
+                    (n == total, lambda: func(*args, **kwargs))),
+                   default=_update_subset)
+
+
+def run_if(cond, func):
```

Comment:
since this "run_if" function is moved from common.py to here. do we need to update the related import where this function is used across alf?

Reply:
It's imported in common.py. So it won't cause any issue.